File: backend/src/main/java/com/predio/mijangos/core/constants/ValidationConstants.java
package com.predio.mijangos.core.constants;

/**
 * Constantes para validación de datos en la aplicación.
 * 
 * <p>Define valores de validación como longitudes mínimas/máximas,
 * patrones regex, y mensajes de error estándar para validaciones.
 * 
 * <p>Estas constantes se usan principalmente en:
 * <ul>
 *   <li>Anotaciones de validación en DTOs (@Size, @Pattern, etc.)</li>
 *   <li>Validadores custom</li>
 *   <li>Validaciones de negocio en servicios</li>
 * </ul>
 * 
 * <p><b>Uso en DTOs:</b>
 * <pre>
 * public record UsuarioCreateDTO(
 *     {@literal @}Size(
 *         min = ValidationConstants.USUARIO_MIN_LENGTH,
 *         max = ValidationConstants.USUARIO_MAX_LENGTH,
 *         message = ValidationConstants.USUARIO_SIZE_MESSAGE
 *     )
 *     String usuario
 * ) {}
 * </pre>
 * 
 * @author Equipo Técnico Predio Mijangos
 * @version 1.0.0
 * @since Octubre 2025
 */
public final class ValidationConstants {

    // ==================== USUARIO ====================

    /**
     * Longitud mínima del código de usuario.
     * Valor: 3 caracteres
     */
    public static final int USUARIO_MIN_LENGTH = 3;

    /**
     * Longitud máxima del código de usuario.
     * Valor: 100 caracteres
     */
    public static final int USUARIO_MAX_LENGTH = 100;

    /**
     * Mensaje de error para validación de tamaño de usuario.
     */
    public static final String USUARIO_SIZE_MESSAGE = 
            "El código de usuario debe tener entre " + USUARIO_MIN_LENGTH + 
            " y " + USUARIO_MAX_LENGTH + " caracteres";

    /**
     * Mensaje de error cuando el usuario es requerido pero está vacío.
     */
    public static final String USUARIO_REQUIRED_MESSAGE = "El código de usuario es obligatorio";

    // ==================== CONTRASEÑA ====================

    /**
     * Longitud mínima de contraseña.
     * Valor: 8 caracteres
     */
    public static final int PASSWORD_MIN_LENGTH = 8;

    /**
     * Longitud máxima de contraseña.
     * Valor: 100 caracteres
     */
    public static final int PASSWORD_MAX_LENGTH = 100;

    /**
     * Patrón regex para validar contraseñas fuertes.
     * Requiere: mayúscula, minúscula, dígito, y al menos 8 caracteres.
     */
    public static final String PASSWORD_PATTERN = "^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d).{8,}$";

    /**
     * Mensaje de error para validación de contraseña.
     */
    public static final String PASSWORD_MESSAGE = 
            "La contraseña debe tener al menos 8 caracteres, incluir mayúsculas, " +
            "minúsculas y números";

    // ==================== NOMBRES Y APELLIDOS ====================

    /**
     * Longitud mínima para nombres.
     * Valor: 2 caracteres
     */
    public static final int NOMBRE_MIN_LENGTH = 2;

    /**
     * Longitud máxima para nombres.
     * Valor: 100 caracteres
     */
    public static final int NOMBRE_MAX_LENGTH = 100;

    /**
     * Mensaje de error para validación de tamaño de nombre.
     */
    public static final String NOMBRE_SIZE_MESSAGE = 
            "El nombre debe tener entre " + NOMBRE_MIN_LENGTH + 
            " y " + NOMBRE_MAX_LENGTH + " caracteres";

    /**
     * Longitud mínima para apellidos.
     * Valor: 2 caracteres
     */
    public static final int APELLIDO_MIN_LENGTH = 2;

    /**
     * Longitud máxima para apellidos.
     * Valor: 100 caracteres
     */
    public static final int APELLIDO_MAX_LENGTH = 100;

    /**
     * Mensaje de error para validación de tamaño de apellido.
     */
    public static final String APELLIDO_SIZE_MESSAGE = 
            "El apellido debe tener entre " + APELLIDO_MIN_LENGTH + 
            " y " + APELLIDO_MAX_LENGTH + " caracteres";

    // ==================== IDENTIFICACIÓN ====================

    /**
     * Longitud mínima de identificación (DPI, NIT, etc.).
     * Valor: 8 caracteres
     */
    public static final int IDENTIFICACION_MIN_LENGTH = 8;

    /**
     * Longitud máxima de identificación.
     * Valor: 20 caracteres
     */
    public static final int IDENTIFICACION_MAX_LENGTH = 20;

    /**
     * Mensaje de error para validación de identificación.
     */
    public static final String IDENTIFICACION_SIZE_MESSAGE = 
            "La identificación debe tener entre " + IDENTIFICACION_MIN_LENGTH + 
            " y " + IDENTIFICACION_MAX_LENGTH + " caracteres";

    /**
     * Patrón para DPI guatemalteco.
     * Formato: 13 dígitos sin espacios ni guiones
     */
    public static final String DPI_PATTERN = "^[0-9]{13}$";

    /**
     * Mensaje de error para DPI inválido.
     */
    public static final String DPI_MESSAGE = "El DPI debe contener exactamente 13 dígitos";

    /**
     * Patrón para NIT guatemalteco.
     * Formato: 7-8 dígitos seguidos de un guión y un dígito verificador
     * Ejemplos: 1234567-8, 12345678-9
     */
    public static final String NIT_PATTERN = "^[0-9]{7,8}-[0-9K]$";

    /**
     * Mensaje de error para NIT inválido.
     */
    public static final String NIT_MESSAGE = 
            "El NIT debe tener el formato correcto (ej: 1234567-8 o 12345678-9)";

    // ==================== TELÉFONO ====================

    /**
     * Longitud mínima de teléfono.
     * Valor: 8 caracteres
     */
    public static final int TELEFONO_MIN_LENGTH = 8;

    /**
     * Longitud máxima de teléfono.
     * Valor: 20 caracteres
     */
    public static final int TELEFONO_MAX_LENGTH = 20;

    /**
     * Patrón para teléfono guatemalteco (móvil y fijo).
     * Formatos aceptados:
     * - 8 dígitos: 12345678
     * - Con código país: +50212345678
     * - Con guiones: 1234-5678
     */
    public static final String TELEFONO_PATTERN = 
            "^(\\+502)?[0-9]{8}$|^[0-9]{4}-[0-9]{4}$";

    /**
     * Mensaje de error para teléfono inválido.
     */
    public static final String TELEFONO_MESSAGE = 
            "El teléfono debe tener formato válido (8 dígitos, ej: 12345678)";

    // ==================== EMAIL ====================

    /**
     * Longitud máxima de email.
     * Valor: 255 caracteres
     */
    public static final int EMAIL_MAX_LENGTH = 255;

    /**
     * Patrón regex para validar email.
     * Más permisivo que el estándar pero funcional para casos comunes.
     */
    public static final String EMAIL_PATTERN = 
            "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$";

    /**
     * Mensaje de error para email inválido.
     */
    public static final String EMAIL_MESSAGE = "El correo electrónico no tiene un formato válido";

    // ==================== DIRECCIÓN ====================

    /**
     * Longitud mínima de dirección.
     * Valor: 10 caracteres
     */
    public static final int DIRECCION_MIN_LENGTH = 10;

    /**
     * Longitud máxima de dirección.
     * Valor: 500 caracteres
     */
    public static final int DIRECCION_MAX_LENGTH = 500;

    /**
     * Mensaje de error para validación de dirección.
     */
    public static final String DIRECCION_SIZE_MESSAGE = 
            "La dirección debe tener entre " + DIRECCION_MIN_LENGTH + 
            " y " + DIRECCION_MAX_LENGTH + " caracteres";

    // ==================== DESCRIPCIÓN ====================

    /**
     * Longitud máxima de descripción.
     * Valor: 1000 caracteres
     */
    public static final int DESCRIPCION_MAX_LENGTH = 1000;

    /**
     * Mensaje de error para descripción muy larga.
     */
    public static final String DESCRIPCION_SIZE_MESSAGE = 
            "La descripción no puede exceder " + DESCRIPCION_MAX_LENGTH + " caracteres";

    // ==================== CÓDIGO/SKU ====================

    /**
     * Longitud mínima de código/SKU.
     * Valor: 3 caracteres
     */
    public static final int CODIGO_MIN_LENGTH = 3;

    /**
     * Longitud máxima de código/SKU.
     * Valor: 50 caracteres
     */
    public static final int CODIGO_MAX_LENGTH = 50;

    /**
     * Patrón para código/SKU alfanumérico.
     * Permite letras, números, guiones y guiones bajos.
     */
    public static final String CODIGO_PATTERN = "^[A-Za-z0-9_-]+$";

    /**
     * Mensaje de error para código inválido.
     */
    public static final String CODIGO_MESSAGE = 
            "El código debe contener solo letras, números, guiones y guiones bajos";

    // ==================== VALORES NUMÉRICOS ====================

    /**
     * Valor mínimo para cantidades.
     * Valor: 0
     */
    public static final int CANTIDAD_MIN = 0;

    /**
     * Valor máximo para cantidades.
     * Valor: 999999
     */
    public static final int CANTIDAD_MAX = 999999;

    /**
     * Mensaje de error para cantidad inválida.
     */
    public static final String CANTIDAD_RANGE_MESSAGE = 
            "La cantidad debe estar entre " + CANTIDAD_MIN + " y " + CANTIDAD_MAX;

    /**
     * Valor mínimo para precios.
     * Valor: 0.01 (un centavo)
     */
    public static final String PRECIO_MIN = "0.01";

    /**
     * Valor máximo para precios.
     * Valor: 9999999.99 (casi 10 millones)
     */
    public static final String PRECIO_MAX = "9999999.99";

    /**
     * Mensaje de error para precio inválido.
     */
    public static final String PRECIO_RANGE_MESSAGE = 
            "El precio debe estar entre " + PRECIO_MIN + " y " + PRECIO_MAX;

    // ==================== PAGINACIÓN ====================

    /**
     * Tamaño mínimo de página.
     * Valor: 1
     */
    public static final int PAGE_SIZE_MIN = 1;

    /**
     * Tamaño máximo de página.
     * Valor: 100
     */
    public static final int PAGE_SIZE_MAX = 100;

    /**
     * Tamaño de página por defecto.
     * Valor: 20
     */
    public static final int PAGE_SIZE_DEFAULT = 20;

    // ==================== MENSAJES GENÉRICOS ====================

    /**
     * Mensaje genérico para campo requerido.
     */
    public static final String FIELD_REQUIRED = "Este campo es obligatorio";

    /**
     * Mensaje genérico para valor inválido.
     */
    public static final String INVALID_VALUE = "El valor proporcionado no es válido";

    /**
     * Mensaje genérico para formato incorrecto.
     */
    public static final String INVALID_FORMAT = "El formato del campo es incorrecto";

    // ==================== CONSTRUCTOR PRIVADO ====================

    /**
     * Constructor privado para prevenir instanciación.
     * Esta es una clase de utilidad que solo contiene constantes estáticas.
     * 
     * @throws UnsupportedOperationException si se intenta instanciar
     */
    private ValidationConstants() {
        throw new UnsupportedOperationException(
                "ValidationConstants es una clase de utilidad y no debe ser instanciada"
        );
    }
}

File: predio-web/src/constants/validation.constants.ts
/**
 * @file validation.constants.ts
 * @description Constantes de validación, patrones regex y límites
 * 
 * Este archivo define:
 * - Patrones regex para validaciones
 * - Límites de longitud para campos
 * - Rangos numéricos permitidos
 * - Mensajes de validación
 * 
 * IMPORTANTE: Debe estar alineado con las validaciones del backend
 * Backend: @Pattern, @Size, @Min, @Max en DTOs
 * 
 * @author Equipo Técnico Predio Mijangos
 * @version 2.0.0
 */

// ============================================
// PATRONES REGEX
// ============================================

/**
 * Patrones de validación regex
 * Sincronizados con las validaciones del backend
 * 
 * @example
 * ```typescript
 * import { VALIDATION_PATTERNS } from '@/constants/validation.constants';
 * 
 * if (!VALIDATION_PATTERNS.EMAIL.test(email)) {
 *   // Email inválido
 * }
 * ```
 */
export const VALIDATION_PATTERNS = {
  /**
   * Username
   * - Solo alfanumérico, punto, guion bajo y guion
   * - 3-50 caracteres
   * Backend: @Pattern(regexp = "^[a-zA-Z0-9._-]+$")
   */
  USERNAME: /^[a-zA-Z0-9._-]+$/,
  
  /**
   * Email
   * RFC 5322 simplificado
   */
  EMAIL: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
  
  /**
   * Contraseña
   * - Al menos 8 caracteres
   * - Al menos una minúscula
   * - Al menos una mayúscula
   * - Al menos un número
   * - Al menos un carácter especial (@$!%*?&#)
   * Backend: @Pattern(regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&#])[A-Za-z\\d@$!%*?&#]{8,}$")
   */
  PASSWORD: /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&#])[A-Za-z\d@$!%*?&#]{8,}$/,
  
  /**
   * Teléfono Guatemala
   * - 8 dígitos
   * - Puede empezar con código de país opcional (+502 o 502)
   * Ejemplos válidos: 12345678, 50212345678, +50212345678
   */
  TELEFONO_GT: /^(\+?502)?[0-9]{8}$/,
  
  /**
   * DPI (Documento Personal de Identificación)
   * - 13 dígitos
   * - Sin espacios ni guiones
   */
  DPI: /^[0-9]{13}$/,
  
  /**
   * NIT (Número de Identificación Tributaria)
   * - Formato: 1234567-8 o 12345678
   * - 7-8 dígitos antes del guion
   * - 1 dígito verificador después del guion (opcional)
   */
  NIT: /^[0-9]{7,8}(-?[0-9kK])?$/,
  
  /**
   * Código postal Guatemala
   * - 5 dígitos
   */
  CODIGO_POSTAL: /^[0-9]{5}$/,
  
  /**
   * Placa de vehículo Guatemala
   * - Formato: P-123ABC o C-123ABC
   * - Letra inicial (P=Particular, C=Comercial, M=Moto, etc.)
   * - Guion
   * - 3 números
   * - 3 letras
   */
  PLACA_VEHICULO: /^[A-Z]-[0-9]{3}[A-Z]{3}$/,
  
  /**
   * Número de serie/VIN
   * - 17 caracteres alfanuméricos
   */
  VIN: /^[A-HJ-NPR-Z0-9]{17}$/,
  
  /**
   * Solo letras y espacios
   * Para nombres, apellidos
   */
  SOLO_LETRAS: /^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s]+$/,
  
  /**
   * Solo números
   */
  SOLO_NUMEROS: /^[0-9]+$/,
  
  /**
   * Alfanumérico
   * Letras, números, espacios
   */
  ALFANUMERICO: /^[a-zA-Z0-9\s]+$/,
  
  /**
   * Código SKU
   * - Alfanumérico con guiones
   * - Ejemplo: PROD-2024-001
   */
  SKU: /^[A-Z0-9-]+$/,
  
  /**
   * URL
   * Validación básica de URL
   */
  URL: /^(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?$/,
  
  /**
   * Hexadecimal
   * Para colores, códigos, etc.
   */
  HEXADECIMAL: /^[0-9A-Fa-f]+$/,
  
} as const;

// ============================================
// LÍMITES DE LONGITUD
// ============================================

/**
 * Límites de longitud para campos de texto
 * Sincronizados con @Size del backend
 * 
 * @example
 * ```typescript
 * import { LENGTH_LIMITS } from '@/constants/validation.constants';
 * 
 * maxLength={LENGTH_LIMITS.USERNAME.MAX}
 * ```
 */
export const LENGTH_LIMITS = {
  // Autenticación y Usuario
  USERNAME: { MIN: 3, MAX: 50 },
  PASSWORD: { MIN: 8, MAX: 100 },
  EMAIL: { MIN: 5, MAX: 100 },
  
  // Nombres
  PRIMER_NOMBRE: { MIN: 2, MAX: 50 },
  SEGUNDO_NOMBRE: { MIN: 0, MAX: 50 },
  PRIMER_APELLIDO: { MIN: 2, MAX: 50 },
  SEGUNDO_APELLIDO: { MIN: 0, MAX: 50 },
  
  // Identificación
  DPI: { EXACT: 13 },
  NIT: { MIN: 8, MAX: 10 },
  TELEFONO: { MIN: 8, MAX: 15 },
  
  // Direcciones
  DIRECCION: { MIN: 10, MAX: 200 },
  CIUDAD: { MIN: 2, MAX: 50 },
  DEPARTAMENTO: { MIN: 2, MAX: 50 },
  CODIGO_POSTAL: { EXACT: 5 },
  
  // Productos
  NOMBRE_PRODUCTO: { MIN: 3, MAX: 100 },
  DESCRIPCION_PRODUCTO: { MIN: 10, MAX: 500 },
  SKU: { MIN: 3, MAX: 50 },
  CODIGO_BARRAS: { MIN: 8, MAX: 20 },
  
  // General
  NOMBRE_CORTO: { MIN: 2, MAX: 50 },
  NOMBRE_LARGO: { MIN: 3, MAX: 100 },
  DESCRIPCION_CORTA: { MIN: 10, MAX: 200 },
  DESCRIPCION_LARGA: { MIN: 20, MAX: 1000 },
  NOTAS: { MIN: 0, MAX: 500 },
  OBSERVACIONES: { MIN: 0, MAX: 1000 },
  
  // Archivos
  NOMBRE_ARCHIVO: { MIN: 1, MAX: 255 },
  
} as const;

// ============================================
// RANGOS NUMÉRICOS
// ============================================

/**
 * Rangos numéricos permitidos
 * Sincronizados con @Min, @Max del backend
 * 
 * @example
 * ```typescript
 * import { NUMERIC_RANGES } from '@/constants/validation.constants';
 * 
 * min={NUMERIC_RANGES.CANTIDAD.MIN}
 * max={NUMERIC_RANGES.CANTIDAD.MAX}
 * ```
 */
export const NUMERIC_RANGES = {
  // Cantidades
  CANTIDAD: { MIN: 0, MAX: 999999 },
  STOCK: { MIN: 0, MAX: 999999 },
  STOCK_MINIMO: { MIN: 0, MAX: 999999 },
  STOCK_MAXIMO: { MIN: 1, MAX: 999999 },
  
  // Precios y montos (en centavos)
  PRECIO: { MIN: 0, MAX: 99999999 }, // Hasta Q999,999.99
  DESCUENTO_PORCENTAJE: { MIN: 0, MAX: 100 },
  DESCUENTO_MONTO: { MIN: 0, MAX: 99999999 },
  
  // Crédito
  CREDITO_DIAS: { MIN: 1, MAX: 365 },
  CREDITO_LIMITE: { MIN: 0, MAX: 9999999999 }, // Hasta Q99,999,999.99
  
  // Paginación
  PAGE_SIZE: { MIN: 1, MAX: 100 },
  PAGE_NUMBER: { MIN: 0, MAX: 999999 },
  
  // Prioridad
  PRIORIDAD: { MIN: 1, MAX: 10 },
  
  // Calificación
  CALIFICACION: { MIN: 1, MAX: 5 },
  
  // Porcentajes
  PORCENTAJE: { MIN: 0, MAX: 100 },
  
  // Edad
  EDAD: { MIN: 18, MAX: 120 },
  
  // Año
  ANIO: { MIN: 1900, MAX: 2100 },
  
} as const;

// ============================================
// TAMAÑOS DE ARCHIVO
// ============================================

/**
 * Tamaños máximos de archivo en bytes
 */
export const FILE_SIZE_LIMITS = {
  /**
   * Imagen de perfil
   * Máximo: 2 MB
   */
  PROFILE_IMAGE: 2 * 1024 * 1024,
  
  /**
   * Imagen de producto
   * Máximo: 5 MB
   */
  PRODUCT_IMAGE: 5 * 1024 * 1024,
  
  /**
   * Documento general
   * Máximo: 10 MB
   */
  DOCUMENT: 10 * 1024 * 1024,
  
  /**
   * Documento grande (reportes, exports)
   * Máximo: 50 MB
   */
  LARGE_DOCUMENT: 50 * 1024 * 1024,
  
} as const;

/**
 * Tipos MIME permitidos por categoría
 */
export const ALLOWED_FILE_TYPES = {
  /**
   * Imágenes
   */
  IMAGES: ['image/jpeg', 'image/png', 'image/gif', 'image/webp'],
  
  /**
   * Documentos
   */
  DOCUMENTS: [
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-excel',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  ],
  
  /**
   * Todos los tipos permitidos
   */
  ALL: [
    'image/jpeg', 'image/png', 'image/gif', 'image/webp',
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-excel',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  ],
  
} as const;

// ============================================
// MENSAJES DE VALIDACIÓN
// ============================================

/**
 * Mensajes de error de validación
 * Usados por Zod y React Hook Form
 * 
 * @example
 * ```typescript
 * import { VALIDATION_MESSAGES } from '@/constants/validation.constants';
 * 
 * email: z.string()
 *   .email(VALIDATION_MESSAGES.EMAIL_INVALID)
 * ```
 */
export const VALIDATION_MESSAGES = {
  // General
  REQUIRED: 'Este campo es obligatorio',
  INVALID_FORMAT: 'El formato no es válido',
  
  // Texto
  TOO_SHORT: (min: number) => `Debe tener al menos ${min} caracteres`,
  TOO_LONG: (max: number) => `Debe tener máximo ${max} caracteres`,
  EXACT_LENGTH: (length: number) => `Debe tener exactamente ${length} caracteres`,
  
  // Números
  TOO_SMALL: (min: number) => `Debe ser mayor o igual a ${min}`,
  TOO_LARGE: (max: number) => `Debe ser menor o igual a ${max}`,
  NOT_NUMBER: 'Debe ser un número',
  NOT_INTEGER: 'Debe ser un número entero',
  NEGATIVE_NUMBER: 'No se permiten números negativos',
  
  // Específicos
  USERNAME_INVALID: 'El usuario solo puede contener letras, números, puntos, guiones bajos y guiones',
  EMAIL_INVALID: 'El correo electrónico no es válido',
  PASSWORD_WEAK: 'La contraseña debe tener al menos 8 caracteres, una mayúscula, una minúscula, un número y un carácter especial',
  PASSWORD_MISMATCH: 'Las contraseñas no coinciden',
  TELEFONO_INVALID: 'El teléfono debe tener 8 dígitos',
  DPI_INVALID: 'El DPI debe tener 13 dígitos',
  NIT_INVALID: 'El NIT no es válido',
  URL_INVALID: 'La URL no es válida',
  
  // Fechas
  DATE_INVALID: 'La fecha no es válida',
  DATE_FUTURE: 'La fecha no puede ser futura',
  DATE_PAST: 'La fecha no puede ser pasada',
  DATE_RANGE_INVALID: 'El rango de fechas no es válido',
  
  // Archivos
  FILE_TOO_LARGE: (maxSize: number) => `El archivo es demasiado grande. Máximo ${formatBytes(maxSize)}`,
  FILE_TYPE_INVALID: 'Tipo de archivo no permitido',
  
  // Comparación
  MUST_MATCH: (field: string) => `Debe coincidir con ${field}`,
  MUST_BE_GREATER: (field: string) => `Debe ser mayor que ${field}`,
  MUST_BE_LESS: (field: string) => `Debe ser menor que ${field}`,
  
  // Selección
  SELECT_REQUIRED: 'Debe seleccionar una opción',
  SELECT_MIN_OPTIONS: (min: number) => `Debe seleccionar al menos ${min} opciones`,
  SELECT_MAX_OPTIONS: (max: number) => `Debe seleccionar máximo ${max} opciones`,
  
} as const;

// ============================================
// HELPERS
// ============================================

/**
 * Formatea bytes a formato legible
 * 
 * @param bytes - Cantidad de bytes
 * @param decimals - Número de decimales
 * @returns String formateado
 * 
 * @example
 * ```typescript
 * formatBytes(1024); // "1 KB"
 * formatBytes(1048576); // "1 MB"
 * ```
 */
export function formatBytes(bytes: number, decimals: number = 2): string {
  if (bytes === 0) return '0 Bytes';
  
  const k = 1024;
  const dm = decimals < 0 ? 0 : decimals;
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
  
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
}

/**
 * Valida si un string cumple con un patrón regex
 * 
 * @param value - Valor a validar
 * @param pattern - Patrón regex
 * @returns true si cumple, false si no
 * 
 * @example
 * ```typescript
 * isValid('test@example.com', VALIDATION_PATTERNS.EMAIL); // true
 * ```
 */
export function isValid(value: string, pattern: RegExp): boolean {
  return pattern.test(value);
}

/**
 * Valida si una longitud está dentro del rango
 * 
 * @param value - Valor a validar
 * @param limits - Límites de longitud
 * @returns true si está dentro del rango
 * 
 * @example
 * ```typescript
 * isLengthValid('test', LENGTH_LIMITS.USERNAME); // false (muy corto)
 * ```
 */
export function isLengthValid(
  value: string,
  limits: { MIN?: number; MAX?: number; EXACT?: number }
): boolean {
  const length = value.length;
  
  if (limits.EXACT !== undefined) {
    return length === limits.EXACT;
  }
  
  if (limits.MIN !== undefined && length < limits.MIN) {
    return false;
  }
  
  if (limits.MAX !== undefined && length > limits.MAX) {
    return false;
  }
  
  return true;
}

/**
 * Valida si un número está dentro del rango
 * 
 * @param value - Valor a validar
 * @param range - Rango numérico
 * @returns true si está dentro del rango
 * 
 * @example
 * ```typescript
 * isInRange(50, NUMERIC_RANGES.DESCUENTO_PORCENTAJE); // true
 * ```
 */
export function isInRange(
  value: number,
  range: { MIN: number; MAX: number }
): boolean {
  return value >= range.MIN && value <= range.MAX;
}

/**
 * Valida si el tipo de archivo es permitido
 * 
 * @param fileType - Tipo MIME del archivo
 * @param allowedTypes - Array de tipos permitidos
 * @returns true si el tipo es permitido
 * 
 * @example
 * ```typescript
 * isFileTypeAllowed('image/jpeg', ALLOWED_FILE_TYPES.IMAGES); // true
 * ```
 */
export function isFileTypeAllowed(
  fileType: string,
  allowedTypes: readonly string[]
): boolean {
  return allowedTypes.includes(fileType);
}

/**
 * Valida si el tamaño del archivo está dentro del límite
 * 
 * @param fileSize - Tamaño del archivo en bytes
 * @param maxSize - Tamaño máximo permitido en bytes
 * @returns true si el tamaño es válido
 * 
 * @example
 * ```typescript
 * isFileSizeValid(1024000, FILE_SIZE_LIMITS.PROFILE_IMAGE); // true
 * ```
 */
export function isFileSizeValid(fileSize: number, maxSize: number): boolean {
  return fileSize <= maxSize;
}